### DISCUSSION TITLE
Validation using Enums is impractical. Can we use sets?

### DISCUSSION THREAD

**Discussion by SvenPVoigt** [pydantic/pydantic#2585] [Question] [answered]
I have an application where I need to accept a value that is part of an allowed set (ie validate that a given value is part of a domain).

I have the following validation cases:

```python
"a" in {"a", "b", "c", "d"}
```

I know this can be handled using Enum objects, but it is a roundabout way and has some unfortunate consequences.

For example, consider the following class:

```python
from enum import Enum
from pydantic import BaseModel, Any
from typing import Dict

allKeys = "a b c d".split(" ")
allKeys = Enum("allKeys", zip(allKeys, allKeys))

class Test(BaseModel):
    the_map: Dict[allKeys, Any]
```

Then the problem is that adding the value `the_map` to the Test class is different than retrieving it.

```python
Test(the_map={"a": 1, "b": 2}).the_map[allKeys.a] # this works
Test(the_map={"a": 1, "b": 2}).the_map["a"] # this does not work
```

Retrieving requires knowledge of the Enum object while adding does not... what if the allKeys object is a hidden part of the module? Then the end user has no way of retrieving the item at that key.

Ideally, I would not have to create an Enum object at all. I would simply validate keys against a set and then be able to access the values again by the same key that was passed to the Test class.

Is this a feature that is being worked on? Is there a way I can contribute to add this feature?

Thanks!

**Comment by PrettyWood** [comment]
Hi @SvenPVoigt
I'm sorry but IMO it's quite logical and pythonic.
If you use a regular enum, it makes sense that keys are enum members.
If you want to also use their representation as valid access, you should inherit from `str` as well so in your case
```py
allKeys = Enum("allKeys", zip(allKeys, allKeys), type=str)
```
Hope it helps!

**Reply by SvenPVoigt** [reply]
Thanks for the reply @PrettyWood! This a good point and I was not aware of this!

I still am not certain Enums are as expressive as Sets. How would it work for mixed types? If I had the dictionary: 

```python
the_map = {"a": "a", 1: 1, ("a", 1): ("a", 1)}
```

I could validate as follows:
```python
assert set( the_map.keys() ).issubset( {"a", 1, ("a", 1)} )
```

And access elements as follows:
```python
the_map[("a", 1)]
```

Would this be possible with Enums?

Thanks!

**Comment by PrettyWood** [comment, marked as answer]
You can do it! An enum needs string as key but you can set a dumb one and use values in pydantic models :)
```py
from enum import Enum
from typing import Any, Dict

from pydantic import BaseModel, ValidationError

valid_keys = {"a", 1, ("a", 1)}
allKeys = Enum("allKeys", {str(i): x for i, x in enumerate(list(valid_keys))})

class Test(BaseModel):
    the_map: Dict[allKeys, Any]

    class Config:
        use_enum_values = True

assert Test(the_map={'a': 1}).the_map['a'] == 1
assert Test(the_map={('a', 1): 2}).the_map[('a', 1)] == 2
try:
    Test(the_map={'b': 1})
    assert False
except ValidationError:
    assert True
```
Hope it helps :)

**Reply by SvenPVoigt** [reply]
That works for me! I think using values in the pydantic models is very useful. Cool!

**Reply by PrettyWood** [reply]
Could you mark the discussion as answered please?
